server/api/users.js

```diff
@@ -311,29 +353,57 @@ router.post('/users/login', async (req, res, next) => {
     let wallet;
     const { platform } = req.body;
 
-    if (!platform) throw new ValidationError('INVALID_PLATFORM');
+    switch (platform) {
+      case 'wallet': {
+        const {
+          from,
+          payload: stringPayload,
+          sign,
+        } = req.body;
+        wallet = from;
+        const isLogin = true;
+        checkSignPayload(wallet, stringPayload, sign, isLogin);
+        const query = await dbRef.where('wallet', '==', wallet).get();
+        if (query.docs.length > 0) {
+          user = query.docs[0].id;
+        }
+        break;
+      }
 
-    if (platform === 'wallet') {
-      const {
-        from,
-        payload: stringPayload,
-        sign,
-      } = req.body;
-      wallet = from;
-      const isLogin = true;
-      checkSignPayload(wallet, stringPayload, sign, isLogin);
-      const query = await dbRef.where('wallet', '==', wallet).get();
-      if (query.docs.length > 0) {
-        user = query.docs[0].id;
+      case 'email':
+      case 'google':
+      case 'twitter': {
+        const { firebaseIdToken } = req.body;
+        const { uid: firebaseUserId } = await admin.auth().verifyIdToken(firebaseIdToken);
+        const query = await dbRef.where('firebaseUserId', '==', firebaseUserId).get();
+        if (query.docs.length > 0) {
+          user = query.docs[0].id;
+        }
+        break;
       }
-    } else {
-      const { firebaseIdToken } = req.body;
-      const { uid: firebaseUserId } = await admin.auth().verifyIdToken(firebaseIdToken);
-      const query = await dbRef.where('firebaseUserId', '==', firebaseUserId).get();
-      if (query.docs.length > 0) {
-        user = query.docs[0].id;
+
+      case 'facebook': {
+        try {
+          const { accessToken, platformUserId } = req.body;
+          await fetchFacebookUser(accessToken);
+          const query = (
+            await oAuthDbRef
+              .where(`${platform}UserId`, '==', platformUserId)
+              .get()
```

Comment:
probably want to add `.limit(1)` for query that only checks for existence
```suggestion
              .get().limit(1)
```
(although this part was originally wrote by me and idk why I didn't add `limit()`..)